Feature 27540 add unicredit bank as new provider

DavidPrexta opened the pull request:
https://is.roihunter.com/issues/27540
Added new provider Unicredit.
@business-factory/pythonistas Please review and merge.

harpener commented:
@DavidPrexta About the situation I described above, we need to use custom `User-Agent` header. 🙄

harpener commented on gold_digger/data_providers/unicredit.py:
```diff
@@ -0,0 +1,123 @@
+        :rtype: requests.Response | None
+        """
+        date_str = date_of_exchange.strftime("%Y%m%dT00:00:00.000+0000")
+        data = {"Currency": "*ALL", "DateFrom": date_str, "DateTo": date_str}
```
@DavidPrexta It's not gonna support **CZK** currency, because it is the base currency in their API. Maybe we can handle it in code? Or is it unnecessary/impossible?

DavidPrexta commented:
@harpener Thanks for the review. I run the code before, it worked fine. But now I got the same error as you so I guess they changed something. I added User-Agent `ROI Hunter /Currency Downloader; http://www.roihunter.com/` (not sure if OK, I got inspired by lady).
I added CZK currency manually with rate 1 so it can now be used too. 
But in doing so I have noticed some weird behavior with the calculation of currency in method `_conversion_to_base_currency`. When converting from `USD` to `CZK` I should get the same rate as in https://www.unicreditbank.cz/cs/obcane/historical.html, so 22.83 but I got 0.043. I found out that problem is that we should calculate `base_currency_rate/currency_rate` in that method to get rate correctly.
I also tried it with other conversions like `USD` to `EUR` where I got rate 1.123..., which is not correct. After altering the method, I got the correct rate of 0.89. Checked it with Google which was 0.88 so almost the same. Same goes for other ratings.
I checked how it works with fixed and I think it should also be switched. Having response form fixed e.g.
```python
{
    "success": true,
    "historical": true,
    "date": "2013-12-24",
    "timestamp": 1387929599,
    "base": "GBP",
    "rates": {
        "USD": 1.636492,
        "EUR": 1.196476,
    }
}
```
where we want to convert `USD` to `EUR` we would calculate it by dividing 1.196476 / 1.196476 which would give 0.7311. Switching it we would get 1.3677 which is almost the same as the rate was for `USD/EUR` on that day https://www.poundsterlinglive.com/best-exchange-rates/euro-to-us-dollar-exchange-rate-on-2013-12-24.
Please check my changes. Maybe I did the math wrong, but rates are corresponding to what they should be.

harpener commented:
@DavidPrexta I saw the changes you made to the rate conversion and I tried to investigate whether it was a bug before or not. I actually came to a conclusion that you fixed a bug, because we use the same calculation on more places and it is always like `to_rate / from_rate`, i.e. `base_currency_rate / currency_rate`. Calling the exchange rate manager for converting EUR to CZK confirmed it. What do you say? :-)

harpener commented:
Hmm, I honestly admit that I am not sure and it's hard to say, because I don't know if exchange rate is how much of `to_currency_rate` makes one `from_currency_rate`. I'm losing the abstract insight. :-) Anyway, keep in mind that some rates returned from Unicredit are 0 and you get division error during the conversion - maybe ignore those rates.

harpener commented on gold_digger/data_providers/unicredit.py:
```diff
@@ -0,0 +1,132 @@
+        "EntityCode": "CZ",
+        "Language": "CS",
+        "SourceSystem": "PWS",
+        "User-Agent": "ROI Hunter /Currency Downloader; http://www.roihunter.com/"
```
I think the user agent is fine, maybe use https instead of http :-)

harpener commented:
Ok, I compared the results from Grandtrunk and Unicredit and came to a conclusion that you did find a bug and it was wrong before in Fixed and it is wrong now that you reverted the fix. :-) You can also write tests for the provider, if you want to. :-)

harpener commented on gold_digger/data_providers/unicredit.py:
```diff
@@ -0,0 +1,132 @@
+    def _get_rates_from_unicredit(self, date_of_exchange, logger):
+        """
+        :type date_of_exchange: datetime.date
+        :param logger: gold_digger.utils.ContextLogger
```
@DavidPrexta Small detail :-)
```suggestion
        :type logger: gold_digger.utils.ContextLogger
```
